bundles/framework/tools.vitruv.framework.correspondence/src/tools/vitruv/framework/correspondence/impl/InternalCorrespondenceModelImpl.xtend

```
@@ -253,240 +55,194 @@ class InternalCorrespondenceModelImpl extends ModelInstance implements InternalC
+		return markedCorrespondences
+	}
 
-	/**
-	 * Removes the current entries in the
-	 * {@link CorrespondenceModelImpl#tuid2CorrespondencesMap} map for the given oldTuid
-	 * before the hash code of it is updated and returns a pair containing the oldTuid and
-	 * the removed correspondence model elements of the map.
-	 * 
-	 * @param oldCurrentTuid
-	 * @return oldCurrentTuidAndStringAndMapEntriesTriple
-	 */
-	override performPreAction(Tuid oldCurrentTuid) {
-		if (tuidUpdateData !== null) {
-			throw new IllegalStateException("Two update calls were running at the same time");
+	private def void markCorrespondenceAndDependingCorrespondencesRecursively(Set<Correspondence> markedCorrespondences,
+		Correspondence correspondence) {
+		markedCorrespondences.add(correspondence)
+		// FIXME MK detect dependency cycles in correspondences already when the reference is updated
+		for (dependingCorrespondence : correspondence.dependedOnBy) {
+			markCorrespondenceAndDependingCorrespondencesRecursively(markedCorrespondences, dependingCorrespondence)
 		}
-		// The Tuid is used as key in this map. Therefore the entry has to be removed before
-		// the hashCode of the Tuid changes.
-		// remove the old map entries for the tuid before its hashcode changes
-		val oldTuidLists = tuid2tuidListsMap.remove(oldCurrentTuid) ?: new HashSet<List<Tuid>>()
-		val oldTuidList2Correspondences = new ArrayList<Pair<List<Tuid>, Set<Correspondence>>>(oldTuidLists.size);
-		for (oldTuidList : oldTuidLists) {
-			val correspondencesForOldTuidList = tuid2CorrespondencesMap.remove(oldTuidList) ?:
-				new HashSet<Correspondence>()
-			oldTuidList2Correspondences.add(
-				new Pair<List<Tuid>, Set<Correspondence>>(oldTuidList, correspondencesForOldTuidList))
+	}
+	
+	private def haveUuids(List<EObject> eObjects) {
+		if (eObjects.forall[uuidResolver.hasPotentiallyCachedUuid(it)]) {
+			return true
+		} else {
+			logger.warn("UUID resolver has no UUID for one of the elements: " + eObjects)
```

Can we make this an exception now? As far as I can see, this is an erroneous state that point to a bug and continuing makes no sense.

---

If so, I guess we can just remove this method because `getPotentiallyCachedUuid` will throw the exception for us, won’t it?

---

This should throw an exception and we should remove the method as soon as possible to have an exception thrown.
However, there are currently some application in which this warning occurs without a negative effect, which would fail upon the change. And even worse, the framework tests throws this warning thousands of times (see #426).
It's on my list to have the bugs in the applications and the framework tests fixed, so that this warning does not occur anymore, such that we can safely remove it.

---

Ah, that’s what I feared. Then we’ll have to leave it in for now. Good that you are on it!

---

I've opened #434 for this issue.